pkg/metrics/metrics.go

```diff
@@ -0,0 +1,67 @@
+/*
+Copyright 2018 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	dto "github.com/prometheus/client_model/go"
+
+	"k8s.io/kube-state-metrics/pkg/options"
+)
+
+type GathererFunc func() ([]*dto.MetricFamily, error)
+
+func (f GathererFunc) Gather() ([]*dto.MetricFamily, error) {
+	return f()
+}
+
+func FilteredGatherer(r prometheus.Gatherer, whitelist options.MetricSet, blacklist options.MetricSet) prometheus.Gatherer {
+	whitelistEnabled := !whitelist.IsEmpty()
+	blacklistEnabled := !blacklist.IsEmpty()
+	filterEnabled := whitelistEnabled || blacklistEnabled
+
+	return GathererFunc(func() ([]*dto.MetricFamily, error) {
```

Comment:
I suggest we split the `GathererFunc` into a inter-package level function, just like `FilteredGatherer` And add some ut for this.

Reply:
@brancz Once you have time, could you please update this PR? :)

Reply:
Can you explain what you meant me to do with `GathererFunc`? It's just so a function satisfies the `prometheus.Gatherer` interface.